game.py

```diff
@@ -42,131 +86,136 @@ def get_sequences(self,c=None):
-
         self.tokens = 0
         self.taken.append(card)
         return (card, tokens)
-
-    def has_cards(self):
-        return len(self.deck) > 0
-    
-    def no_thanks(self):
-        self.tokens += 1
     
-
-
 class Game:
-
-    def __init__(self, n_players=4):
-
-        # Player Setup
-        if 3 > n_players > 7:
-            return Error("Player count must be between 3 and 7 inclusive. You provided {}".format(n_players))
-        
-        self.n_players = n_players
-        self.players = []
-        ptot = {
-            3:11,
-            4:11,
-            5:11,
-            6:9,
-            7:7
-        }
-
-        for i in range(0, self.n_players):
-            self.players.append(Player(ptot[n_players], i))
-        
-        # deck setup
+    """
+    Class for the Game of No Thanks, contains all of the rules and actions
+    """
+
+    def __init__(self, players=[]):
+        # Make a list of all the players
+        self.players = players
+        # Count the number of players
+        self.n_players = len(self.players)
+        # Deck setup
         self.deck = Deck()
-        self.current_player_index = 0
+        # Set up Turn Counter
         self.turn_counter = 0
 
-        self.game_log = []
-
-    def get_turn_index(self):
+        # Deal tokens to all the players
+        n_tokens = {3:11, 4:11, 5:11, 6:9, 7:7}
+        for player in self.players:
+            player.tokens = n_tokens[self.n_players]
+                
+    @property
+    def turn(self):
         return self.turn_counter%self.n_players
 
-    def get_state(self):
-        player_scores = [(p.player_number, p.calc_score()) for p in self.players]
-        player_scores.sort(key=lambda x : x[1])
-
-        state = {
-            "flipped_card" : self.deck.get_flipped(),
+    @property
+    def state(self):
+        d = {
+            "flipped_card" : self.deck.flipped_card,
             "tokens_on_card": self.deck.tokens,
-            "player_states" : [p.get_state() for p in self.players],
-            "player_positions" : [p[0] for p in player_scores],
-            "player_turn_index" : self.get_turn_index()
+            "player_states" : [p.state for p in self.players],
+            "player_turn_index" : self.turn
         }
-        return state
+        return d
     
-    def player_action(self, choice):
-        player = self.players[self.get_turn_index()]
-        old_score = player.calc_score()
-        if choice >= 0:
+    def player_action(self):
+        """
+        Function to allow the player to take an action
+
+        Parameters
+        ----------
+        choice: Boolean
+            If True, the player takes the card.
+            If False, the player says "No Thanks!"
+        """
+
+        # Get the player whose turn it is
+        player = self.players[self.turn]
+        # Ask the player what they want to do
+        choice = player.decide(self.state)
+
+        # The player decides to take the card
+        if choice:
```

Comment:
it may be a good idea to check if player's choice is valid, e.i, checking it has the token needed to skip.